sklearn/metrics/_classification.py

```diff
@@ -547,47 +555,52 @@ def multilabel_confusion_matrix(
 
         # labels are now from 0 to len(labels) - 1 -> use bincount
         tp = y_true == y_pred
-        tp_bins = y_true[tp]
+        tp_indices = xp.nonzero(tp)[0]
+        tp_bins = xp.reshape(xp.take(y_true, tp_indices, axis=0), (-1,))
         if sample_weight is not None:
-            tp_bins_weights = np.asarray(sample_weight)[tp]
+            tp_bins_weights = xp.take(xp.asarray(sample_weight), tp_indices, axis=0)
```

Comment:
We might need to be specific about the device here as well.

Reply:
According to the ArrayAPI spec https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html
The dtype and device are both inferred from the initial array. Since sample_weight is already set to the corresponding device in our tests
```python
    if metric_kwargs.get("sample_weight") is not None:
        metric_kwargs["sample_weight"] = xp.asarray(
            metric_kwargs["sample_weight"], device=device
        )
```
I think that is why no error is being raised.